### ISSUE TITLE
[LethalEnergy] bug report + question

### ISSUE THREAD

**Issue by JohnMetroid** [Miloet/TheManyModsOfLethalCompany#7]
hey there,

I spent some time testing lethal energy and it's a really cool mod.  I'm getting the following error when removing cans from the case:

```
[09:47:44.8124244] [Message:Mellowdy.LethalEnergy] CansHeld: 6
[09:47:44.8134244] [Message:Mellowdy.LethalEnergy] SetCans was called AND went through with 5 cans
[09:47:44.8144257] [Error  : Unity Log] SerializationException: [Network] Parameter (Vector3 vector) is not marked [Serializable] nor does it implement INetworkSerializable!
Stack trace:
RuntimeNetcodeRPCValidator.FastBufferExtensions.WriteMethodInfoAndParameters (Unity.Netcode.FastBufferWriter fastBufferWriter, System.Reflection.MethodBase methodInfo, System.Object[] args) (at <0df0f351a5d94cbb9af1af6290fbdcfa>:IL_00CF)
RuntimeNetcodeRPCValidator.NetworkBehaviourExtensions.MethodPatchInternal (Unity.Netcode.NetworkBehaviour networkBehaviour, System.Reflection.MethodBase method, System.Object[] args) (at <0df0f351a5d94cbb9af1af6290fbdcfa>:IL_00A5)
RuntimeNetcodeRPCValidator.NetworkBehaviourExtensions.MethodPatch (Unity.Netcode.NetworkBehaviour __instance, System.Reflection.MethodBase __originalMethod, System.Object[] __args) (at <0df0f351a5d94cbb9af1af6290fbdcfa>:IL_0000)
(wrapper dynamic-method) LC_LethalEnergy.CanHandler.DMD<LC_LethalEnergy.CanHandler::SpawnCanClientRpc>(LC_LethalEnergy.CanHandler,Unity.Netcode.NetworkObjectReference,UnityEngine.Vector3,int)
(wrapper dynamic-method) LC_LethalEnergy.CanHandler.DMD<LC_LethalEnergy.CanHandler::SpawnCanServerRpc>(LC_LethalEnergy.CanHandler)
System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <787acc3c9a4c471ba7d971300105af24>:IL_006A)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at <787acc3c9a4c471ba7d971300105af24>:IL_0083)
System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) (at <787acc3c9a4c471ba7d971300105af24>:IL_0000)
RuntimeNetcodeRPCValidator.NetworkBehaviourExtensions.ReceiveNetworkMessage (System.UInt64 sender, Unity.Netcode.FastBufferReader reader) (at <0df0f351a5d94cbb9af1af6290fbdcfa>:IL_0143)
Unity.Netcode.CustomMessagingManager.InvokeNamedMessage (System.UInt64 hash, System.UInt64 sender, Unity.Netcode.FastBufferReader reader, System.Int32 serializedHeaderSize) (at <895801699cfc4b4ab52267f31e2a4998>:IL_00CB)
Unity.Netcode.CustomMessagingManager.SendNamedMessage (System.String messageName, System.UInt64 clientId, Unity.Netcode.FastBufferWriter messageStream, Unity.Netcode.NetworkDelivery networkDelivery) (at <895801699cfc4b4ab52267f31e2a4998>:IL_0060)
RuntimeNetcodeRPCValidator.NetworkBehaviourExtensions.MethodPatchInternal (Unity.Netcode.NetworkBehaviour networkBehaviour, System.Reflection.MethodBase method, System.Object[] args) (at <0df0f351a5d94cbb9af1af6290fbdcfa>:IL_00F7)
RuntimeNetcodeRPCValidator.NetworkBehaviourExtensions.MethodPatch (Unity.Netcode.NetworkBehaviour __instance, System.Reflection.MethodBase __originalMethod, System.Object[] __args) (at <0df0f351a5d94cbb9af1af6290fbdcfa>:IL_0000)
(wrapper dynamic-method) LC_LethalEnergy.CanHandler.DMD<LC_LethalEnergy.CanHandler::SpawnCanServerRpc>(LC_LethalEnergy.CanHandler)
LC_LethalEnergy.DrinkCase.ItemActivate (System.Boolean used, System.Boolean buttonDown) (at <3cd7c37a8a7f4821bef69ad4f53ecb51>:IL_0066)
GrabbableObject.UseItemOnClient (System.Boolean buttonDown) (at <255c0877f9414c3aa6f1fe5079ac13f3>:IL_005C)
(wrapper dynamic-method) GameNetcodeStuff.PlayerControllerB.DMD<GameNetcodeStuff.PlayerControllerB::ActivateItem_performed>(GameNetcodeStuff.PlayerControllerB,UnityEngine.InputSystem.InputAction/CallbackContext)
UnityEngine.InputSystem.Utilities.DelegateHelpers.InvokeCallbacksSafe[TValue] (UnityEngine.InputSystem.Utilities.CallbackArray`1[System.Action`1[TValue]]& callbacks, TValue argument, System.String callbackName, System.Object context) (at <dbb0ce7b96144bd8bcbcb8b0d765f970>:IL_001B)
UnityEngine.Logger:LogException(Exception, Object)
AsyncLoggers.Wrappers.Unity.<>c__DisplayClass4_0:<LogException>b__0() (at ./src/Wrappers/Unity/AsyncLoggerWrapper.cs:36)
AsyncLoggers.Wrappers.Unity.<>c__DisplayClass28_0:<wrapCallback>b__0() (at ./src/Wrappers/Unity/AsyncLoggerWrapper.cs:136)
AsyncLoggers.Wrappers.JobWrapper:LogWorker() (at ./src/Wrappers/JobWrapper.cs:84)
AsyncLoggers.Wrappers.LogJob:Execute() (at ./src/Wrappers/JobWrapper.cs:68)
Unity.Jobs.JobStruct`1:Execute(LogJob&, IntPtr, IntPtr, JobRanges&, Int32)

[09:47:44.8154258] [Error  : Unity Log] TargetInvocationException while executing 'performed' callbacks of 'Movement/ActivateItem[/Mouse/leftButton,/XInputControllerWindows/rightTrigger]'
```
It doesn't seem to cause issues with the can dropping.

I also wanted to ask, do energy drinks currently only spawn on the moons listed in the config?  Would it be possible to get an option to set a global spawn weight that's applicable to all moons (or an option to specify a list of weights attached to each moons (e.g. Experimentation:12,Assurance:20,Vow:30,March:40,Offense:10)

I think the scrap values are a little bit weird, too.  For instance, a drink case with 6 cans was worth 9 credits.  After dropping a can: 59c case + 8c drink.  After dropping another can: 49c case + 2x 8c drink.  Then 39c case and 3x 8c drinks.

**Comment by Miloet** [comment]
I believe I know roughly what is causing the error and will try to fix it.

They do only spawn in the listed moons I believe, I can make the default weight be applicable to all moons outside the base ones (excluding the new ones as they are not yet included in the API from my knowledge)
> Would it be possible to get an option to set a global spawn weight that's applicable to all moons (or an option to specify a list of weights attached to each moons (e.g. Experimentation:12,Assurance:20,Vow:30,March:40,Offense:10)

Do you mean that instead of having a base value, such as 20, and a multiplier for each moon, such as 1.5, resulting in 30 you think it would be better to have each moon have an assigned weight directly?

It being worth 9 credits inditially is not intended, however the dropped can prices are correct. Case price + 10c per can. The logic goes that selling it with all the cans in it thus the player has an insentive to keep the cans in the case when selling.
I could possibly add a config option for this however that sets the price for the can, the case and the case scaling with the can. That way its up to the player. While I am at it I could add config for the speedboost.

**Comment by JohnMetroid** [comment]
Thanks for pushing an update, I'm going to dive into it later tonight hopefully :)

>Do you mean that instead of having a base value, such as 20, and a multiplier for each moon, such as 1.5, resulting in 30 you think it would be better to have each moon have an assigned weight directly?

Yeah, my preference would be to either set a truly global spawn rate **or** directly set the spawn weight for each moon.  Any method to get the loot spawning globally, really.

**Comment by Miloet** [comment]
I did implement that! However you light want to delete the config file so
that none of the old setting stay there. Changing setting names,
descriptions and values can create ”phantom settings” that dont actually do
anything. Deleting it and starting the game however should do the trick.

**Comment by JohnMetroid** [comment]
I got some time to check the new update and everything looks great! Reported error for can removal is fixed and new config is perfect, thanks so much.

One more thing I noticed.  When holding an individual can (but not a case), the console is spammed with tons of updates about the player holding the can (I think?).  With async loggers it probably won't cause performance issues, but figured I'd bring to your attention

![image](https://github.com/Miloet/TheManyModsOfLethalCompany/assets/139426529/9cc026ed-220e-4070-9714-559baa24916e)

thanks again for the update, really enjoying the mod :)
